File: web/src/main/java/com/canyon/web/RequestHandler.java
package com.canyon.web;

import com.canyon.commons.JsonUtils;
import com.canyon.commons.StringUtils;
import com.canyon.core.TypeRef;
import com.canyon.web.parser.HandlerValueParser;
import com.canyon.web.router.WebRouter;
import com.canyon.web.router.WebRouterParam;
import io.vertx.core.Handler;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.ext.web.FileUpload;
import io.vertx.ext.web.RoutingContext;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

public class RequestHandler implements Handler<RoutingContext> {
    private WebRouter webRouter;
    private HandlerValueParser handlerValueParser;
    private WebParameterParser webParameterParser;

    public RequestHandler(WebRouter webRouter, HandlerValueParser handlerValueParser, WebParameterParser webParameterParser) {
        this.webRouter = webRouter;
        this.handlerValueParser = handlerValueParser;
        this.webParameterParser = webParameterParser;
    }

    @Override
    public void handle(RoutingContext rc) {
        HttpServerResponse response = rc.response();
        response.setChunked(true);
        Method method = webRouter.getMethod();
        Object[] parameters = new Object[webRouter.getWebParams().size()];
        for (int i = 0; i < parameters.length; i++) {
            WebRouterParam webParam = webRouter.getWebParams().get(i);
            Class<?> clazz = webParam.getClazz();
            if (clazz.equals(HttpServerRequest.class)) {
                parameters[i] = rc.request();
            } else if (clazz.equals(HttpServerResponse.class)) {
                parameters[i] = rc.response();
            } else if (clazz.equals(new TypeRef<List<FileUpload>>() {
            }.getType())) {
                if (StringUtils.isNotEmpty(webParam.getName())) {
                    List<FileUpload> lists = new ArrayList<>();
                    for (FileUpload file : rc.fileUploads()) {
                        if (webParam.getName().equals(file.name())) {
                            lists.add(file);
                        }
                    }
                    parameters[i] = lists;
                } else {
                    parameters[i] = new ArrayList<>(rc.fileUploads());
                }
            } else if (clazz.equals(RoutingContext.class)) {
                parameters[i] = rc;
            } else {
                String value = webParameterParser.parser(webParam, rc);
                parameters[i] = handlerValueParser.parser(webParam, value);
            }
        }

        try {
            Object retObj = method.invoke(webRouter.getController(), parameters);
            if (method.getReturnType() != null) {
                response.putHeader("content-type", webRouter.getResponse());
                if (retObj != null) {
                    if (MediaType.APPLICATION_JSON.equals(webRouter.getResponse())) {
                        response.write(JsonUtils.toString(retObj));
                    } else {
                        response.write(retObj.toString());
                    }
                }
            }
            response.end();
        } catch (IllegalAccessException e) {
            rc.fail(e);
        } catch (InvocationTargetException e) {
            rc.fail(e);
        }
    }
}

File: web/src/main/kotlin/com/canyon/web/RequestHandler.kt
package com.canyon.web

import com.canyon.commons.JsonUtils
import com.canyon.core.TypeRef
import com.canyon.inject.Bean
import io.vertx.core.Handler
import io.vertx.core.http.HttpServerRequest
import io.vertx.core.http.HttpServerResponse
import io.vertx.ext.web.FileUpload
import io.vertx.ext.web.RoutingContext
import kotlin.reflect.full.createType

@Bean
class RequestHandler(
        private val webRouter: WebRouter,
        private val handlerValueParser: HandlerValueParser,
        private val webParameterParser: WebParameterParser
) : Handler<RoutingContext> {
    override fun handle(rc: RoutingContext) {
        val response = rc.response()
        response.isChunked = true
        val func = webRouter.function
        val parameters = arrayOfNulls<Any?>(webRouter.webParam.size)
        for (i in 0 until parameters.size) {
            val webParam = webRouter.webParam[i]
            webParam.kType == object : TypeRef<List<FileUpload>> {}.kType
            when {
                webParam.kType.classifier == HttpServerRequest::class -> parameters[i] = rc.request()
                webParam.kType.classifier == HttpServerResponse::class -> parameters[i] = rc.response()
                webParam.kType == object : TypeRef<List<FileUpload>> {}.kType -> {
                    if(webParam.name.isNotEmpty()) {
                        val files = mutableListOf<FileUpload>()
                        rc.fileUploads().forEach {
                            if (it.name() == webParam.name) {
                                files.add((it))
                            }
                        }
                        parameters[i] = files
                    }else{
                        parameters[i] = rc.fileUploads().toList()
                    }
                }
                webParam.kType.classifier == RoutingContext::class -> parameters[i] = rc
                else -> {
                    val strValue = webParameterParser.parser(webParam, rc)
                    parameters[i] = handlerValueParser.parser(webParam, strValue)
                }
            }
        }
        val ret = func.call(webRouter.controller, *parameters)
        if (func.returnType != Unit::class && ret != null) {
            response.putHeader("content-type", webRouter.responseMediaType.toText())
            if (ret !== Unit) {
                writeResponse(response, webRouter.responseMediaType, ret)
            }
        }
        response.end()
    }

    private fun writeResponse(response: HttpServerResponse, mediaType: MediaType, obj: Any?) {
        if (obj == null)
            return
        when (mediaType) {
            MediaType.APPLICATION_JSON -> response.end(JsonUtils.toString(obj))
            else -> response.end(obj.toString())
        }
    }
}
